### ISSUE TITLE
String leak with allocated zero-length c_string

### ISSUE THREAD

**Issue by benharsh** [chapel-lang/chapel#7914]
### Summary of Problem
Strings can be constructed from c_strings. A common pattern is to specify ``needToCopy=false``, which allows the string to take ownership of the c_string's memory. This is useful when dealing with extern functions that allocate strings.

The string will not take ownership of the memory if the c_string's length is zero or if the pointer is nil. This means that even if the c_string was allocated, the string will leak if for whatever reason the first byte is the null terminator.

Here is an example:
```chpl

config const zero = false;

proc main() {
  var data = c_calloc(uint(8), 5);
  if zero == false then data[0] = 65;
  var cs = data:c_string;
  var s = new string(cs, owned=true, needToCopy=false);

  writeln("s = '", s, "'");
}
```
```
> ./a.out --memLeaks
s = 'A'

=================================================================================================================
Allocated Memory (Bytes)         Number   Size     Total    Description                      Address             
=================================================================================================================
=================================================================================================================
> ./a.out --memLeaks --zero
s = ''

============================================================================================================================================
Allocated Memory (Bytes)                                    Number   Size     Total    Description                      Address             
============================================================================================================================================
$CHPL_HOME/modules/internal/LocaleModelHelpMem.chpl:67      1        5        5        array elements                   0x00007fa5dd402cb0  
============================================================================================================================================
```

**Comment by mppf** [comment]
Right, I think it'd be reasonable to change reinitString (around line 261) to only check for buf == nil, since it might have to delete buf if s_len == 0 buf buf is not nil.

**Comment by cassella** [comment]
This is the generalization of #7549, right?

**Comment by benharsh** [comment]
Perhaps, but it looked like there was a separate problem in the issue you linked, just in the same function.

**Comment by benharsh** [comment]
Resolved by #7918.
